File: src/main/java/play/pluv/login/application/JwtProvider.java
package play.pluv.login.application;

import static io.jsonwebtoken.io.Decoders.BASE64;

import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import java.time.Duration;
import java.util.Date;
import java.util.Map;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class JwtProvider {

  public static final String MEMBER_ID_IDENTIFIER = "memberId";
  private static final String TOKEN_ISSUER = "pluv";
  private static final int DURATION_DAY = 90;
  private final String encodedSecretKey;

  public JwtProvider(@Value("${jwt.secret}") final String encodedSecretKey) {
    this.encodedSecretKey = encodedSecretKey;
  }

  public String createAccessTokenWith(final Long memberId) {
    final Date now = new Date();
    final Date expiration
        = new Date(now.getTime() + Duration.ofDays(DURATION_DAY).toMillis());
    final Map<String, Object> claims = Map.of(MEMBER_ID_IDENTIFIER, memberId);

    return Jwts.builder()
        .issuer(TOKEN_ISSUER)
        .issuedAt(now)
        .expiration(expiration)
        .claims(claims)
        .signWith(Keys.hmacShaKeyFor(BASE64.decode(encodedSecretKey)))
        .compact();
  }

  public Long parseMemberId(final String jwt) {
    final Integer memberId = (Integer) Jwts.parser()
        .verifyWith(Keys.hmacShaKeyFor(BASE64.decode(encodedSecretKey)))
        .build()
        .parseSignedClaims(jwt)
        .getPayload()
        .get(MEMBER_ID_IDENTIFIER);
    return Long.valueOf(memberId);
  }
}

File: src/main/kotlin/play/pluv/security/JwtProvider.kt
package play.pluv.security

import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders.BASE64
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.*

@Component
class JwtProvider(
    @param:Value($$"${jwt.secret}") private val encodedSecretKey: String
) {

    companion object {
        private const val MEMBER_ID_IDENTIFIER = "memberId"
        private const val TOKEN_ISSUER = "pluv"
        private const val DURATION_DAY = 90
    }

    fun createAccessTokenWith(memberId: Long): String {
        val now = Date()
        val expires = Date(now.time + DURATION_DAY * 1000)
        val claims = mapOf(MEMBER_ID_IDENTIFIER to memberId)

        return Jwts.builder()
            .issuer(TOKEN_ISSUER)
            .issuedAt(now)
            .expiration(expires)
            .claims(claims)
            .signWith(Keys.hmacShaKeyFor(BASE64.decode(encodedSecretKey)))
            .compact()
    }

    fun parseMemberId(jwt: String): Long {
        val memberId = Jwts.parser()
            .verifyWith(Keys.hmacShaKeyFor(BASE64.decode(encodedSecretKey)))
            .build()
            .parseSignedClaims(jwt)
            .payload[MEMBER_ID_IDENTIFIER] as Number

        return memberId.toLong()
    }
}